File: src/main/java/com/duxtinto/intellij/plugin/github/codereviews/di/impl/dagger/modules/CodeReviewModule.java
package com.duxtinto.intellij.plugin.github.codereviews.di.impl.dagger.modules;

import com.duxtinto.intellij.plugin.github.codereviews.data.codereviews.CodeReviewsDataContract;
import com.duxtinto.intellij.plugin.github.codereviews.data.codereviews.CodeReviewsRepository;
import com.duxtinto.intellij.plugin.github.codereviews.di.scopes.ProjectScoped;
import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.reviews.CodeReviewsDomainContract;
import com.duxtinto.intellij.plugin.github.codereviews.net.codereviews.apiv3.CodeReviewsFetcher;
import com.duxtinto.intellij.plugin.github.codereviews.ui.toolwindow.codereviews.CodeReviews;
import com.duxtinto.intellij.plugin.github.codereviews.ui.toolwindow.codereviews.CodeReviewsPresenter;
import com.duxtinto.intellij.plugin.github.codereviews.ui.toolwindow.codereviews.CodeReviewsView;
import dagger.Binds;
import dagger.Module;

@Module
public abstract class CodeReviewModule {
    @Binds
    @ProjectScoped
    public abstract CodeReviewsDomainContract.Repository provideCodeReviewsDomainContractRepository(CodeReviewsRepository repository);

    @Binds
    @ProjectScoped
    public abstract CodeReviewsDataContract.Fetcher provideCodeReviewsDataContractFetcher(CodeReviewsFetcher fetcher);

    @Binds
    @ProjectScoped
    public abstract CodeReviews.View provideCodeReviewsView(CodeReviewsView view);

    @Binds
    @ProjectScoped
    public abstract CodeReviews.Presenter provideCodeReviewsPresenter(CodeReviewsPresenter presenter);
}

File: src/main/kotlin/com/duxtinto/intellij/plugin/github/codereviews/di/impl/dagger/modules/CodeReviewModule.kt
package com.duxtinto.intellij.plugin.github.codereviews.di.impl.dagger.modules

import com.duxtinto.intellij.plugin.github.codereviews.data.codereviews.CodeReviewsDataContract
import com.duxtinto.intellij.plugin.github.codereviews.data.codereviews.CodeReviewsRepository
import com.duxtinto.intellij.plugin.github.codereviews.data.codereviews.comments.CommentsRepository
import com.duxtinto.intellij.plugin.github.codereviews.di.scopes.ProjectScoped
import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.reviews.CodeReviewsDomainContract
import com.duxtinto.intellij.plugin.github.codereviews.net.codereviews.apiv3.CodeReviewsFetcher
import com.duxtinto.intellij.plugin.github.codereviews.presentation.toolwindow.codereviews.CodeReviews
import com.duxtinto.intellij.plugin.github.codereviews.presentation.toolwindow.codereviews.CodeReviewsPresenter
import com.duxtinto.intellij.plugin.github.codereviews.presentation.toolwindow.codereviews.CodeReviewsTreeCellRenderer
import com.duxtinto.intellij.plugin.github.codereviews.presentation.toolwindow.codereviews.CodeReviewsView
import com.duxtinto.intellij.plugin.github.codereviews.presentation.toolwindow.codereviews.events.MouseListener
import dagger.Binds
import dagger.Module
import javax.inject.Named
import javax.swing.tree.TreeCellRenderer

@Module(includes = [
    ReviewCommentModule::class
])
abstract class CodeReviewModule {
    @Binds
    @ProjectScoped
    abstract fun provideCodeReviewsDomainContractRepository(repository: CodeReviewsRepository): CodeReviewsDomainContract.Repository

    @Binds
    @ProjectScoped
    abstract fun provideCodeReviewCommentsDomainContractRepository(repository: CommentsRepository): CodeReviewsDomainContract.Comments.Repository

    @Binds
    @ProjectScoped
    abstract fun provideCodeReviewsDataContractFetcher(fetcher: CodeReviewsFetcher): CodeReviewsDataContract.Fetcher

    @Binds
    @ProjectScoped
    abstract fun provideCodeReviewsView(view: CodeReviewsView): CodeReviews.View

    @Binds
    @ProjectScoped
    @Named("code_reviews")
    abstract fun provideCodeReviewsTreeCellRenderer(cellRenderer: CodeReviewsTreeCellRenderer): TreeCellRenderer

    @Binds
    @ProjectScoped
    abstract fun provideCodeReviewsViewEventsMouseListener(listener: MouseListener): CodeReviews.View.Events.MouseListener

    @Binds
    @ProjectScoped
    abstract fun provideCodeReviewsPresenter(presenter: CodeReviewsPresenter): CodeReviews.Presenter
}
